File: 1-easy/problems/0501-find-mode-in-binary-search-tree/solution.cpp
class Solution {
public:
    vector<int> findMode(TreeNode* root) {
        unordered_map<int, int> counts;
        dfs(root, counts);

        int maxCount = 0;
        for (const auto& [value, count] : counts) {
            maxCount = max(maxCount, count);
        }

        vector<int> result;
        for (const auto& [value, count] : counts) {
            if (count == maxCount) {
                result.push_back(value);
            }
        }
        return result;
    }

private:
    void dfs(TreeNode* node, unordered_map<int, int>& counts) {
        if (!node) {
            return;
        }
        counts[node->val]++;
        dfs(node->left, counts);
        dfs(node->right, counts);
    }
};

File: 1-easy/problems/0501-find-mode-in-binary-search-tree/solution.py
class Solution:
    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        counts = {}

        def dfs(node):
            if not node:
                return
            counts[node.val] = counts.get(node.val, 0) + 1
            dfs(node.left)
            dfs(node.right)

        dfs(root)
        max_count = max(counts.values())
        return [value for value, count in counts.items() if count == max_count]
